standardized-metadata-nested.py

```diff
@@ -0,0 +1,225 @@
+
+
+class SnowflakeTableMetadata(Metadata):
+    """Metadata fields that apply to assets that are tables in Snowflake."""
+
+    snowflake_address: Optional[SnowflakeTableAddress]
+    cluster_by: Optional[str]
+
+    @classmethod
+    def namespace(cls) -> str:
+        return "dagster_snowflake"
+
+    @classmethod
+    def version(cls) -> str:
+        return "1.0.0"
+
+
+####################################################################################################
+# An asset with metadata constructed using the structured metadata APIs.
+####################################################################################################
+
+
+@asset
+def asset1():
+    return MaterializeResult(
+        metadata={
+            **AssetMetadata(
+                storage_kind="snowflake", storage_address_string="my_db.my_schema.asset1"
+            ),
+            **TableMaterializationMetadata(
+                num_rows_total=500,
+                num_rows_inserted=5,
+                columns=[TableColumn(name="user_id", data_type="str")],
+            ),
+            **SnowflakeTableMetadata(
+                snowflake_address=SnowflakeTableAddress(
+                    database="my_db", db_schema="my_schema", table_name="asset1"
+                )
+            ),
+            # non-structured metadata
+            "my_unschematized_piece_of_metadata": 5,
+            "staging_address": SnowflakeTableAddress(
+                database="my_staging_db", db_schema="my_schema", table_name="asset1"
+            ).dict(),
+        }
+    )
+
+
+####################################################################################################
+# An asset with metadata constructed using plain JSON-objects. Has identical metadata to the asset
+# above.
+####################################################################################################
+
+
+@asset
+def asset2():
+    return MaterializeResult(
+        metadata={
+            "dagster.table": {
+                "_version": "1.0.0",
+                "num_rows_inserted": 5,
+                "columns": [{"name": "user_id", "data_type": "str"}],
+                "num_rows_total": 500,
+            },
+            "dagster": {
+                "_version": "1.0.0",
+                "storage_kind": "snowflake",
+                "storage_address_string": "my_db.my_schema.asset1",
+            },
+            "dagster_snowflake": {
+                "_version": "1.0.0",
+                "snowflake_address": {
+                    "database": "my_db",
+                    "db_schema": "my_schema",
+                    "table_name": "asset1",
+                },
+            },
+            # non-structured metadata
+            "my_unschematized_piece_of_metadata": 5,
+            "staging_address": {
+                "database": "my_staging_db",
+                "db_schema": "my_schema",
+                "table_name": "asset1",
+            },
+        }
+    )
```

Comment:
Definitely prefer this to the special treatment of the top-level keys in option one. I think there are two issues here:

1) There is no schema to avoid naming collisions with top-level fields past, present, and future. I still like we should hide of all of these underneath some sort of special key.
2). Do we want to be able to place multiple instances of metadata blobs? This would mean the option to "name" them.

Reply:
> Do we want to be able to place multiple instances of metadata blobs? This would mean the option to "name" them.

Thinking aloud here:

To make this concrete, let's imagine a situation where someone would want the Snowflake "cluster_by" key ("cluster_by" is an attribute of a Snowflake table) to show up multiple times within the metadata blob for a single AssetMaterialization.

So to back up, why would they want to do this? The main reason I can think of would be because there are multiple Snowflake tables that are relevant to a particular materialization. For example, there's the Snowflake table that's the asset itself that's being materialized, and then there's also a "staging" Snowflake table that was somehow produced during the process of materializing the asset.

So assuming we want to record some Snowflake-specific attributes about this staging table, it's possible that we'd also want to log some other metadata fields that are relevant to tables, but not just Snowflake tables, e.g. "num_rows_total".

So, at least naively, I would imagine the materialization-level metadata blob should end up containing some sub-blob that represents this staging table and contains the set of attributes that describe it.

In the unnested variant, this would look like:
```json
{
    "dagster_snowflake/cluster_by": "user_name",
    "staging_table": {
        "dagster_snowflake/cluster_by": "user_id",
        "dagster.table/num_rows_total": 5
    }
}
```

In the nested variant, this would look like:
```json
{
    "__namespaced_metadata": {
        "dagster_snowflake": {"cluster_by": "user_id"}
    }
    "staging_table": {
        "__namespaced_metadata": {
            "dagster_snowflake": {"cluster_by": "user_id"},
            "dagster.table": {"num_rows_total": 5}
        }
    }
}
```